src/base/ui-extensions/ui-extension.ts

```diff
@@ -0,0 +1,120 @@
+/********************************************************************************
+ * Copyright (c) 2019 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { inject, injectable } from "inversify";
+import { ILogger } from "../../utils/logging";
+import { CommandExecutionContext } from "../commands/command";
+import { TYPES } from "../types";
+import { ViewerOptions } from "../views/viewer-options";
+
+/**
+ * A UI extension displaying additional UI elements on top of a sprotty diagram.
+ */
+export interface IUIExtension {
+    readonly id: string;
+    show(context: CommandExecutionContext): void;
+    hide(): void;
+}
+
+/**
+ * Abstract base class for UI extensions.
+ */
+@injectable()
+export abstract class AbstractUIExtension implements IUIExtension {
+
+    abstract readonly id: string;
+    abstract readonly containerClass: string;
+    protected containerElement: HTMLElement;
+    protected activeElement: Element | null;
+
+    constructor(
+        @inject(TYPES.ViewerOptions) protected options: ViewerOptions,
+        @inject(TYPES.ILogger) protected logger: ILogger) { }
+
+    show(context: CommandExecutionContext): void {
```

Comment:
It should be sufficient to pass the root element (maybe marked with `Readonly<>` to indicate that it shouldn't be modified here). The CommandExecutionContext is related to command execution and viewer updates, and should not be accessed in this context.